File: Army/ability/ability.cpp
#include "ability.h"

Ability::Ability(Unit* currentUnit) : currentUnit(currentUnit) {}

Ability::~Ability() {}

void Ability::attack(Unit* victim) {
    victim->takeDamage(currentUnit->getDamage());
    victim->counterAttack(currentUnit);
}

void Ability::counterAttack(Unit* victim) {
    currentUnit->takeDamage(victim->getDamage() / 2);
    
    if ( victim->getUnitType() == vampire ) {
        victim->getCurrentState()->vampirism(currentUnit);
    }
}

void Ability::takeMagicDamage(int damage) {
    currentUnit->getCurrentState()->takeMagicDamage(damage);
}

void Ability::castSpell(Unit* victim, Spell* spell) {}

void Ability::changeState() {
    State* currentState = currentUnit->getCurrentState();
    State* nextState = currentUnit->getNextState();
    State* tmp = currentState;

    currentUnit->setCurrentState(nextState);
    currentUnit->setNextState(tmp);
    currentState = nextState;

    int newCurrentHP = (getHealthMultiplier() * (double)currentState->getHPLimit());

    currentUnit->setHPLimit(currentState->getHPLimit());
    currentUnit->setCurrentHP(newCurrentHP); 
    currentUnit->setDamage(currentState->getDamage());
}

double Ability::getHealthMultiplier() const {
    return (double)currentUnit->getCurrentHP() / (double)currentUnit->getHPLimit();
}

File: Java/src/com/gymfox/army/ability/Ability.java
package com.gymfox.army.Ability;

import com.gymfox.army.Units.Unit;

public abstract class Ability {
    private Unit currentUnit;

    public Ability(Unit currentUnit) {
        this.currentUnit = currentUnit;
    }

    public void attack(Unit victim) throws Unit.UnitIsDeadException, Unit.IsSelfAttackException {
        victim.takeDamage(currentUnit.getDamage());
        victim.counterAttack(currentUnit);
    }

    public void counterAttack(Unit victim) throws Unit.UnitIsDeadException {
        victim.takeDamage(getCurrentUnit().getDamage() / 2);
    }

    public void takeDamage(int damage) throws Unit.UnitIsDeadException {
        tryToKill(damage);
    }

    public void takeMagicDamage(int damage) throws Unit.UnitIsDeadException {
        tryToKill(damage);
    }

    private void tryToKill(int damage) throws Unit.UnitIsDeadException {
        if ( currentUnit.getCurrentHP() <= damage ) {
            currentUnit.setCurrentHP(0);
            currentUnit.notifyObservers();
            currentUnit.notifyObservable();

            return;
        }

        currentUnit.setCurrentHP(currentUnit.getCurrentHP() - damage);
    }

    public void changeState() {}

    public Unit getCurrentUnit() {
        return currentUnit;
    }
}